### PR TITLE
Add minetest particle overlays  (weather incoming ? :))

### PR DESCRIPTION
**Pull request by nerzhul**
Particle overlays are a on-screen particle generator, designed to render many textures affected by gravity and directions on screen. It's very lightweight and permits things like doing snow, rain, leaves (fall season ?)

I added /test_weather command on the experimental game to have a demo.

New video: https://www.youtube.com/watch?v=jTHQQf3auhI

### CODE REVIEW

**Comment by rubenwardy** [comment]
This feels wrong to me. The rendering is the harder part, and this probably is too specific an API to expose.

In anycase, it would probably be better to work out how to do the rendering efficiently before making an API, as that would inform its design

**Comment by nerzhul** [comment]
@rubenwardy let me finish the PoC please :)
i'm working on the rendering part, plugged in the renderer and client environment, and you will see the design is quite nice

**Comment by nerzhul** [comment]
And now you have a basic rain & snow handling, other cases are currently not handled

**Comment by paramat** [comment]
Interesting experiment :)
What is the reason for a specific weather communication? My impression was that the implementation of weather would be done completely by mods using mod channels once server-sent CSM is possible? 

My concern is that this PR approach would pre-determine some aspects of weather instead of leaving it completely to mods. MT is universal so assumptions about having wind, rain and snow are already limitations and might not be suitable for alien planets or fantasy worlds. Also specific and unforseeable weather types like duststorms etc. would then be left out.

**Comment by nerzhul** [comment]
@paramat weather is currently implement in SSM and it's just bad. Exposing renderer to CSM is really a bad idea for those particle generators, it can be implemented but too heavy.
Weather is a generic idea. But in reality you will not have many cases, maybe some parts can be more generic but it's really fine tuning on the client code, i'm not sure we should let modders modify some fine calculations i added to the renderer, the global function attributes i added should be sufficient. 
your point is, i think, to have more generic weather types, it can be possible, for example to have a generic rain + snow, maybe, but currently i'm showing the possibilities. Maybe we can have embedded types, like rain & snow, and let mods register some custom weather types with fine tuning later ?
I don't want to make this PR too huge, i think i will let it in this state, handling snow, rain & wind, make it merged, and then we can think about enhancing it by implementing more types, or modding types for modders

**Comment by rubenwardy** [comment]
I don't really like this, as I feel like this could be implemented as part of the particle API - ie: a specific particle draw type which works well for whether. This also stops this being hard coded.

Moving weather to CSM should make much difference really, the current issue is with the particle implementation - a scene node is used per particle which makes it incredibly inefficient!

I'll have to check out this PR though, the rendering parts could still be useful

**Comment by nerzhul** [comment]
please test it, and review the code. I stop the PR as is, i hope we can implement it as is (or with minor modification). Extensibility can be possible in future PR

**Comment by nerzhul** [comment]
particle API is too low level, this API is much higher and near the renderer. The only acceptable part we can expose is the renderWeatherParticles part which more high level, and it's exposed though the weather API. Many calculation & algorithms are linked with renderer directly

**Comment by rubenwardy** [comment]
I think that weather is too high level. I'd prefer to have an "omnipresent" particle type specific to one player, using the existing API to track them maybe. So to mods, there's no real difference. Also means we don't force the concept of weather just different rendering types which could be useful. Client-side, this could use a totally different rendering method which could be a good idea when you contrast single particles versus uncontrollable multitude of particles.

Tried it out there are a few bugs such as visible starting locations and unanimated direction changes but I like it. Could use with seeing higher densities of particles.

Got this error after a while:


```lua
2018-07-01 18:17:04: ERROR[Server]: /home/ruben/dev/minetest/src/weather.cpp:27: void Weather::State::setType(const string&): A fatal error occured: Invalid weather type set: 'storm'
```

**Comment by nerzhul** [comment]
oops i disabled storm but forgot to remove the lua call :)

**Comment by rubenwardy** [comment]
Relevant PR: #6540

**Comment by paramat** [comment]
I agree with rubenwardy. It's good if you just want to experiment with this in MT without intending to merge, but this is the wrong approach for MT because it is so specific and hardcoded. This is more suited for a hardcoded game like Voxelands.

Server-sent CSM, better particles and other universal features is what will improve weather. The correct approach is to provide the universal non-specific features that enable weather to be implemented by mods, using mod channels between server and client mods to communicate about weather.

You may just want to experiment and share the experiment with others by making a PR, which is fine, but in terms of merging into MT it is best that i state now that i am strongly opposed, sorry, no offense :) I still really appreciate that you are working on MT.

**Comment by nerzhul** [comment]
closing. I lost my day.

**Comment by rubenwardy** [comment]
You could still use this PR as a way to test and make algorithms to cull particles when indoors, for future use. As well as proving that Irrlicht particles is up to the job. So not all is lost

Also note my concerns where mainly API related

**Comment by nerzhul** [comment]
reopening, i will see if i can transform it, later, in a generic thing, or it's not possible

**Comment by paramat** [comment]
The TODO looks interesting, hopefully a sort of a flexible 'mass atmospheric particles' API. Definable velocity direction would be good for sideways stuff like dust storms etc. This would have many uses outside of weather too.
> node over head protection should disable them while walking (check that every 250 to 500 ms over 100 nodes)

Checking that light is 15 is the best way to detect 'under open sky', as we have caves many hundreds of nodes tall (and could be bigger in new or customised mapgens).

I still get a feeling that server-sent CSM should be implemented first as it is so closely related to this intention, and will probably alter and reduce what is needed here. However i do not want to discourage you from experimenting now if you are inspired to do so.

**Comment by paramat** [comment]
Actually i retract that about SSCSM, a better particle API capable of weather is so useful it seems good to have without requiring CSMods.

**Comment by paramat** [comment]
I feel this needs some discussion. My initial thoughts:

This is so closely related to particles i think this should be considered an extension of the particle API (even though it won't create the particles in the same way) and not mention 'weather' at all. Mentioning weather is too specific whereas particles are more generic and multipurpose.

We have 'add particle' and 'add particle spawner', this could be a 3rd implementation something like 'add particle environment', with a similar set of parameters. Similar to a particle spawner but a moving particle spawning box that is always centred on the camera, but with particle velocity and acceleration relative to world co-ordinates.

Particles really need improvement and if this work can also be used to reduce the FPS impact of all particles that would be excellent. So my suggestion is to approach this as part of work to improve the particles API and nor consider it any sort of new or separate API.

About your video:

When view pitches up and down do your particle textures stay vertical?
I assume the particles have a fixed location in the world so that when when the player moves the particles don't shift with the player.
(A note about rain and snow: In your video they both spawn stationary and accelerate downwards, maybe that was just your initial work, but rain and snow are at a constant terminal velocity as they fall.)

**Comment by nerzhul** [comment]
The PSSN storage is now generic

**Comment by nerzhul** [comment]
@paramat to answer abouti verticality, i didn't found how to edit the PSSN texture for currently rendered textures. Maybe we need to patch the irrlicht base class

**Comment by nerzhul** [comment]
this PR is now generic and ready for a review & a merge

**Comment by paramat** [comment]
Some comments and questions:

Texture needs a size parameter like the particle API.

'Intensity' being 0.1 to 10 is abstract and limited to whatever limits you chose. An unlimited absolute parameter like 'particles spawned per second', like the particle API, would be better.

Having 'Wind direction' and 'Wind speed' is not generic and only applies a velocity in the XZ axes, how is Y velocity set?

Instead we could have a v3f 'Velocity' like the particle API.
Direction in degrees and speed is less useful than a velocity vector in nodes per second, as mod code usually works with velocity vectors.
'Wind speed' is also limited to what you have chosen to be the limit, i don't think there should be a limit.

'Gravity factor' seems to be missing from the lua_api.txt docs. I assume it applies a downwards Y acceleration that is the currently set global gravity * gravity factor, so '1' means -9.81? Does this set a specific acceleration in nodes per second squared or is it an abstract value?

However 'gravity' is also not generic, as this may be used in a space game with no gravity. More flexible would be a v3f 'Acceleration' as in the particle API, that then also makes X, positive Y and Z accelerations possible and is independent of the globally set gravity.

We could have parameters for the box containing the random particle initial positions as in the particle API: minp and maxp relative to the player position.

```
+		em = pssn->createBoxEmitter(
+			aabb3f(-100.0f, -130.0f, -100.0f, 110.0f, 200.0f, 110.0f),
```
I assume this is the box around the player that has particles? I think the radius of this box needs to have a parameter as it will affect FPS. This box size in the particle API is critical to tune.

Essentially, i think this should be more similar to the particle spawner API with similar parameter types all in units of seconds and ndes, as this is obviously a particle API. As it is now it seems too specific to a limited idea of certain types of weather. It's not generic and not very flexible.
Perhaps i misunderstand, is it even possible for these particles to 'exist' in the world and have velocity and acceleration vectors aligned to world axes?

A minor note about the test snow: Snow falls at constant speed as it falls at terminal velocity, so there should be no acceleration of gravity (sorry to be pedantic, i have a physics degree).

**Comment by paramat** [comment]
It's not ready for a merge. Tested, some further comments:

The particles have a hardcoded lifetime, there needs to be a parameter for particle lifetime.
The particles are spawned at random locations within the entire box around the player, instead of spawning at a required location set by parameters (such as above the player) and travelling past.
As a result of the above 2 points, slow moving snowflakes appear at random locations, move for a couple of seconds then disappear, looks very weird.

I noticed that particles stop spawning every time i look up above a pitch of roughly 45 degrees.

Each particle has it's vertical size squashed by an amount according to camera angle at particle spawn. When panning the camera upwards this causes particles to have varying vertical height:

screenshot_20180710_001502

This screenshot is looking straight up. Some particles are edge-on as they should be, but most are still seen from the side with varying vertical squash. This will be very noticeable with slow moving and long-lifetime particles.

There is no Y velocity parameter, particles are all spawned stationary at random points around the player and then accelerate under a variable gravity parameter which is used to affect the particle speed.
Instead, IRL rain and snow falls at a constant terminal velocity with no acceleration. In most weather situations a constant velocity is desired with no acceleration, but there is no way to set a velocity.

If gravity is included here it should be set to the physics override value of gravity for the player, not be a parameter here. However, better would be to have a v3f acceleration vector as in the particle API.

There is no way to have particles travelling upwards if desired, for example a biome with steam or smoke rising from the surface. I tried setting gravity to 0 and particles then have 0 Y velocity.

**Comment by paramat** [comment]
The vertical squash for particles needs to be optional, as snowflakes should appear the same at all camera angles. Similar to the 'vertical' parameter in the particle API.

Particles spawn when the player is underwater, i can see this is coded to avoid that so probably just a bug.

Particles still spawn when the player is undercover, however i know the light == 15 mehtod has many issues.
Particles having no collision detection is not a problem, i wouldn't expect that seeing how high-performance this is, but that then means particles fly through the insides of structures.

Particles are fullbright in dark places and at night. Having node light affect each particle may be difficult though and possibly impossible wit this approach. But it's a big problem.

The approach of this PR is still hardcoded and limited, it's similar to the first version and i have the same concerns. By MTEngine standards it's very hardcoded and 'faked', as in, the particles are quite detached from the world they are in and don't act like textures in a world.
I feel it's very unsuitable and out of character for Minetest, even if some of the issues i have mentioned are fixed. It's more of an interesting personal experiment.
The visual and behavioural quality of the implementation is low even if the code is high quality.

It seems too much has been compromised in order to allow such a high particle count. I think it is better to work on the existing particle API and improve that (such as working on the 'one scene node per particle' issue or implementing Irrlicht particles) even if it means a lower particle count than this is capable of.
I don't know why this is so different from the particle API and so inflexible relative to that, with much fewer and very different types of parameter.
:-1:

**Comment by nerzhul** [comment]
@paramat i suggest you to update your comments basing them on particle scene nodes, as some are not possible and only depend on irrlicht methods. I will change the PPS vs intensity which is a facility from my weather first PoC, to restore the PPS.
Here you have 1 scene node per particle overlay, and it's very fast. our current particle system is shit to be honest, and slow. It can only be used for spawners or a single particle spawn on an action. and no i will not work on the current system, as this system solves the main problem i want to address: weather in MT.

**Comment by nerzhul** [comment]
please also note if you see "wind" somewhere it's because i just forget to rename the variable, it's direction & directional speed

**Comment by nerzhul** [comment]
@paramat if you another time block my PR because you want me to work another part which i doesn't want to work on , because it's totally unrelated, this time i leave MT and it will be definitive. it's not the first time you do that shitty behaviour on my PRs or others PRs.

**Comment by nerzhul** [comment]
and you are wrong on rain & snow, gravity affect those particles, just their weight is tiny then the gravity doesn't accelerate many them. When you throw from a plane you will always accelerate due to gravity. It's basic physics. Here you can ajust gravity factor to work on the Y velocity, it's already there

**Comment by ClobberXD** [comment]
The spawning and despawning of the particles are very apparent, as rubenwardy said. It'd be nice if they fade in and fade out instead.

**Comment by nerzhul** [comment]
@ClobberXD if i remember we can apply fading, will look at this later

**Comment by hkzorman** [comment]
I want to chime to show my support. 

While some of paramat's comment in regards to adding more parameters are reasonable, not all of them are. For example, this is a *particle overlay* which means they shouldn't be specific to any player or position in-world. 

I have seen this feature, a particle overlay in many other games. It can be used for many many things, so I think right now this is generic enough to the many uses, which include weather, damage effects, status effects, etc.

Please include this into Minetest. It is a very useful feature.

**Comment by nerzhul** [comment]
thanks @hkzorman , my idea is not to expose irrlicht to server, but a functional element to server, we should NEVER rely server on the graphics engine. I think i added sufficient abstract permitting to remove irrlicht at a point if needed, here.

**Comment by paramat** [comment]
Although this is rendered in a different way, this is a particle API and is an infinite lifespan particle spawner, so i feel it should be similar to the existing particle spawner API in terms of parameters, abstraction and flexibility. Obviously some of the parameters are not possible due to the rendering method.
My suggestion for parameters is at least these (possibly more parameters could have randomisation):
```
    {
        amount = 1,
^ Spawns the amount on a per-second basis.
        radius = 32,
^ Radius of box centred on player within which particles appear.
        minpos = {x=0, y=0, z=0},
        maxpos = {x=0, y=0, z=0},
^ Particles spawn at random positions within this box.
        vel = {x=0, y=0, z=0},
        acc = {x=0, y=0, z=0},
        exptime = 1,
        size = 1,
        vertical = false,
^ vertical: if true faces player using y axis only
        texture = "image.png",
^ Uses texture (string)
        playername = "singleplayer"
^ Playername is optional, if specified spawns particle only on the
^ player's client.
    }
```
This would make the API very flexible and allow many other uses other than rain or snow type weather.

Apart from the parameters and behaviour, my concern about the rendering method is that in order to achieve the very high particle count, which is impressive, too much else is being sacrificed, resulting in the issues i mention.
It would be better to use a rendering method that can't cope with as many particles but where each particle appears and behaves in a higher quality way like the existing particle API. This is what i suspect will be possible if the existing particle API is worked on or a different method is used.

**Comment by paramat** [comment]
> update your comments basing them on particle scene nodes, as some are not possible and only depend on irrlicht methods.

Yes this is what i suspected and i accept some things may not be possible, some of those things are essential, this is part of my concern about the method used.

> i will not work on the current implement system, it's out of this PR scope

Dev work is voluntary of course, however when working on weather the particle API is very much related and in scope. There is obvious work to be done on particles that will certainly improve them, which then affects how weather is done. To clarify i'm not asking you to work on particles, but hopefully someone else will and that should ideally be done before this, as it is likely to provide a better solution for weather.

>  if you see "wind" somewhere it's because i just forget to rename the variable, it's direction & directional speed

Ok :) But direction in degrees and speed is still a horizontal-only velocity addition, so my concern still stands that it is too inflexible and is obviously based on the idea of wind.

> block my PR because you want me to work another part which i doesn't want to work on

You misunderstand, i didn't ask you to work on something else first, i wrote:
> Please can we work on improving existing particles first?

And the particle API is very closely related because this is part of it.
> this time i leave MT and it will be definitive. it's not the first time you do that shitty behaviour on my PRs or others PRs.

I have been unpleasant in the past yes, as have you and almost every core dev.
However in this thread i am reviewing and stating very valid concerns, and expressing my opinion. I have not been particularly unpleasant, the strongest thing i wrote was:
> It's actually not remotely ready for a merge

Which is the simple truth due to the many issues. I was just surprised you considered it mergeable.

I know you are somewhat unhappy with MT for various reasons. I'm sorry that i disagree with you on many aspects of MT, but that's something we all have to cope with, it's nothing personal, just different opinions.

> and you are wrong on rain & snow, gravity affect those particles, just their weight is tiny then the gravity doesn't accelerate many them.

Nope.
Yes gravity affects them, but the force of gravity is equal and opposite to the upwards force of air resistance, at terminal velocity the 2 forces balance and no further acceleration occurs. This occurs long before rain or snow get anywhere near ground level.
You can research this to see if i'm correct.

In this PR velocity cannot be specified by altering gravity because the particles are always accelerating, so constant velocity is impossible.
> Gravity is necessary and handled on the irrlicht side, then you cannot ignore it

Gravity is another word for acceleration, acceleration may be necessary but is zero for particles at terminal velocity and many objects in many situations.

**Comment by nerzhul** [comment]
please note the particle appearance is handled by the API provided in the modding itself, but yes i can improve some points. The API idea you provided has some interesting things to use

**Comment by paramat** [comment]
Actually, it was wrong of me to say i'll disapprove whatever the implementation is, sorry.
My current :-1: is due to the current implementation and the specific issues i mentioned, so let's take it from there.

**Comment by nerzhul** [comment]
changing texture should, i think, use a new pssn (overlay), it's why i choose to use one per weather type in the experimental mod, this permit to handle the texture change glitch, also if you change texture it's for all current particles, then the modder should be careful about that to have a nice rendering

**Comment by hkzorman** [comment]
@nerzhul Can you use more than one particle overlay at the same time? Thinking for something like rain, small variations in the rain drop texture would be good to have.

**Comment by nerzhul** [comment]
i rebased it. If i get time i finish it tomorrow

**Comment by paramat** [comment]
hkzorman, for randomised particle size it's much better to do that within 1 overlay like particle spawner does:
```
        minsize = 1,
        maxsize = 1,
-- ^ The particle's properties are random values in between the bounds:
```
Some other parameters could possibly have randomisation too. It's best that the API is as similar to a particle spawner as possible.

**Comment by nerzhul** [comment]
i will change the parameters to use a table in the next commit i will do, and then i will see what parameters can be like particlespawner or not

**Comment by nerzhul** [comment]
@paramat i also think we cannot use random particle size in this model. I just imported the irrlicht size affector in the project to perming changing particle size for another usage, it affects all particles on screen, like all affectors.

**Comment by nerzhul** [comment]
@paramat , @SmallJoker can you review another time ? i'm interested by your comments. I think we can merge it as is, except if you want some parameters to be renamed.

File: doc/client_lua_api.txt [hunk 1]
```diff
@@ -708,6 +708,10 @@ Call these functions only at load time!
     * Called when the local player open inventory
     * Newest functions are called first
     * If any function returns true, inventory doesn't open
+* `minetest.register_on_particle_overlay_spec(func(name, enabled, minpps, maxpps,
+	direction, speed, gravity_factor))`
```
Comment (SmallJoker) [inline doc/client_lua_api.txt]: Incorrect indents

File: doc/lua_api.txt [hunk 2]
```diff
@@ -4995,6 +4995,19 @@ This is basically a reference to a C++ `ServerActiveObject`
...
+		* `texture_name`: Texture used by particles
+		* `enabled`: Enable or disable the overlay
+		* `minpps`: Minimum particles spawned per second (max `20000`)
+		* `maxpps`: Maximum particles spawned per second (max `20000`)
+		* `direction`: Wind direction (between `0` and `359` degrees)
+		* `velocity`: Particle velocity applied on the defined direction (between `0.0`
+		  and `500.0`)
+		* `gravity_factor`: Physics gravity factor (between `0.0` and `100.0`)
+		* `size_x`: Texture width scale factor (default `1.0`)
+		* `size_y`: Texture height scale factor (default `1.0`)
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Ditto. Tabs are superior, but here we use spaces.
Also: The return value needs documentation

File: src/client/renderingengine.cpp [hunk 3]
```diff
@@ -121,6 +125,146 @@ RenderingEngine::~RenderingEngine()
...
+	}
+
+	em->setMinParticlesPerSecond(spec.minpps);
+	em->setMaxParticlesPerSecond(spec.maxpps);
+
+	if (pssn->getMaterial(0).getTexture(0) != texture)
+		pssn->setMaterialTexture(0, texture);
+
+	for (auto &affector: pssn->getAffectors()) {
+		if (affector->getType() == scene::EPAT_GRAVITY) {
```
Comment (SmallJoker) [inline src/client/renderingengine.cpp]: switch/case

File: src/client/renderingengine.cpp [hunk 4]
```diff
@@ -121,6 +125,146 @@ RenderingEngine::~RenderingEngine()
...
+			addParticleSystemSceneNode(false);
+		pssn->setParent(s_singleton->m_device->getSceneManager()->getActiveCamera());
+		s_singleton->m_particle_overlays[spec.name] = pssn;
+	} else {
+		pssn = overlayIt->second;
+	}
+
+	scene::ICameraSceneNode *camera = s_singleton->m_device->getSceneManager()->
+		getActiveCamera();
+	scene::IParticleEmitter* em = pssn->getEmitter();
```
Comment (SmallJoker) [inline src/client/renderingengine.cpp]: `*em`

File: src/network/clientopcodes.cpp [hunk 5]
```diff
@@ -113,7 +113,8 @@ const ToClientCommandHandler toClientCommandTable[TOCLIENT_NUM_MSG_TYPES] =
 	{ "TOCLIENT_UPDATE_PLAYER_LIST",       TOCLIENT_STATE_CONNECTED, &Client::handleCommand_UpdatePlayerList }, // 0x56
 	{ "TOCLIENT_MODCHANNEL_MSG",           TOCLIENT_STATE_CONNECTED, &Client::handleCommand_ModChannelMsg }, // 0x57
 	{ "TOCLIENT_MODCHANNEL_SIGNAL",        TOCLIENT_STATE_CONNECTED, &Client::handleCommand_ModChannelSignal }, // 0x58
-	null_command_handler,
+	{ "TOCLIENT_PARTICLE_OVERLAY",         TOCLIENT_STATE_CONNECTED,
+		&Client::handleCommand_ParticleOverlay }, // 0x59
```
Comment (SmallJoker) [inline src/network/clientopcodes.cpp]: Technically you're right here, but it's more readable to have them in the same line.

File: src/network/clientpackethandler.cpp [hunk 6]
```diff
@@ -1433,3 +1434,35 @@ void Client::handleCommand_ModChannelSignal(NetworkPacket *pkt)
...
+	ParticleOverlaySpec *spec = m_env.getParticleSpecOverlay(specName);
+	spec->name = specName;
+	*pkt >> spec->texture_name >> spec->texture_scale_factor.Width
+		>> spec->texture_scale_factor.Height >> spec->minpps >> spec->maxpps
+		>> spec->direction >> spec->velocity >> spec->gravity_factor >> spec->enabled;
+
+	// Limit max PPS to prevent bad performance values
+#ifdef __ANDROID__
+	// Android devices are less powerful, lower the limits
+	if (spec->minpps > 5000)
```
Comment (SmallJoker) [inline src/network/clientpackethandler.cpp]: `spec->minpp = rangelim(spec->minpps, 0, 5000);` to ensure that all values are in the range we want them (even though it's sent by the server)

File: src/network/serveropcodes.cpp [hunk 7]
```diff
@@ -202,7 +202,7 @@ const ClientCommandFactory clientCommandFactoryTable[TOCLIENT_NUM_MSG_TYPES] =
 	{ "TOCLIENT_UPDATE_PLAYER_LIST",       0, true }, // 0x56
 	{ "TOCLIENT_MODCHANNEL_MSG",           0, true}, // 0x57
 	{ "TOCLIENT_MODCHANNEL_SIGNAL",        0, true}, // 0x58
-	null_command_factory,
+	{ "TOCLIENT_PARTICLE_OVERLAY",                  0, true}, // 0x59
```
Comment (SmallJoker) [inline src/network/serveropcodes.cpp]: Misaligned indent.

File: src/script/common/helper.cpp [hunk 8]
```diff
@@ -67,6 +68,32 @@ template <> s16 LuaHelper::readParam(lua_State *L, int index)
 	return lua_tonumber(L, index);
 }
 
+template <> u16 LuaHelper::readParam(lua_State *L, int index)
+{
+	return (u16)luaL_checknumber(L, index);
```
Comment (SmallJoker) [inline src/script/common/helper.cpp]: Risk of over/underflows is quite high when reading a double. Needs `rangelim`

**Comment by paramat** [comment]
> i also think we cannot use random particle size in this model.

Ok no problem :) It's not essential.
I'll review again very soon.

**Comment by paramat** [comment]
Almost all of the issues i described above are not attended to. Most have been ignored and not been responded too.
This may be well coded but is very inflexible and causes highly weird particle behaviour that is completely unusable for weather and most other uses. See my other comments above for the details.
With attention to the issues this possibly has a lot of potential.
:-1: sorry.

**Comment by paramat** [comment]
Really has potential, just needs a rework of parameters and behaviour, light at player pos needs to alter particle brightness (i assume per-node light cannot be used due to the high performance of this), then there's the inherently difficult issue of whether to stop the particles spawning when the player is 'undercover' (light level < 15) because the particles come straight through nodes.

In testing i notice that particle textures are squashed vertically according to camera angle as before, but not so much so that they are still visible when looking straight upwards. This still needs to be optional as it does not apply to snow (and will not be needed for many effects) which looks the same viewed from every direction, it also needs to be optional because it looks weird. I'm wondering if this vertical squash should happen at all, but without it raindrop textures would have to be round. It's a difficult issue to resolve.

Haing minpps and maxpps is not useful as the variation occurs over very short time scales so they average out to the average value and have no visible effect. Most effects, including weather, tend to have a constant number of visible particles.

I notice now that particles still appear from nowhere but 'fade-in' by increasing texture size from 0. This has a strange visual effect that makes them look like they are moving towards the player as they fade in. There is no 'fade-out', they just disappear (not that i'm asking for a 'fade-out'). If particles are spawned in a distant defined box (as they need to for weather effects) then the 'fade-in' is not needed as the spawn points will be distant. This 'fading-in' isn't needed.

All other issues are as described earlier in the thread. needs a lot of rework and is not mergeable until we get close to something reasonable.

**Comment by rubenwardy** [comment]
I'd quite like the opinion of someone experienced with graphics on this, ie: an actual graphics programmer

**Comment by paramat** [comment]
These are what i consider the minimum required parameters, and the best form (similar to particle spawner because the feature is):
```
    {
        amount = 1,
^ Spawns the amount on a per-second basis.
        overlay_radius = 32,
^ Radius of box centred on player within which particles are visible.
        spawn_minpos = {x=0, y=0, z=0},
        spawn_maxpos = {x=0, y=0, z=0},
^ Particles spawn at random positions within this volume.
        vel = {x=0, y=0, z=0},
        acc = {x=0, y=0, z=0},
        exptime = 1,
        size = 1,
        collisiondetection = false, (? if possible)
^ collisiondetection: if true uses collision detection
        collision_removal = false, (? if possible or maybe always-on?)
^ collision_removal: if true then particle is removed when it collides,
^ requires collisiondetection = true to have any effect
        texture = "image.png",
^ Uses texture (string)
        playername = "singleplayer"
^ Playername is optional, if specified spawns particle only on the
^ player's client.
        brightness = 0
^ Optional, specify particle brightness. If left out particle brightness
^ Is determined by time of day. Values 0-14.
    }
```
Concerning whether to stop particles when undercover or underwater, i think this can be decided by mod code as the desired behaviours will greatly vary.

Concerning brightness, particles are always fullbright, even at night. I assume it is not possible for every particle to take on the light of the node it is currently passing through? Or is it possible but intensive? Can each particle have a different brightness or all the same?

But anyway, i suggest a 'brightness' parameter to define brightness if a mod wants, if left out the particle brightness is determined by the time of day: so the particle brightness matches the brightness of a node that is open to sky.

For weather, collision detection will be extremely useful, otherwise there will be weird behaviouor with particles coming through roofs or walls, is it possible to have collision detection as an option, even if it reduces performance? It will be well worth losing some performance to gain collision detection as your code can cope with an extreme amount of particles.

I wrote the 'snowdrift' mod https://forum.minetest.net/viewtopic.php?f=11&t=6854 which is considered the best weather mod MT has (note the particles spawn in a thin box high above and fall with constant velocity, please try it and see how much better it looks. I also studied the terminal velocities of different raindrop and snowflake sizes) but the issues of particles coming through roofs and walls means i consider collision detection almost essential.

> 11:11 nerzhul         i think we can merge it in its current state. If we need more and more parameters later we can add it later
11:11 nerzhul         in the current state it solves many usecases and permits to implement weather without problem

No, and no not remotely. It needs a rework before it can be merged due to the many issues, which seem to be intentionally ignored.

I'm concerned that, if merged as it is, the required changes will then simply be ignored and refused such that it remains in code but in a bad state. I'm a little concerned that the issues i raised are being ignored and the questions i asked are not being answered.
It seems possible i will be ignored by the author when they feel like it, however, i'm an expert in physics and in how to implement weather in MT, and have tested carefully and gone into a lot of detail.

**Comment by Wuzzy2** [comment]
I did not test this. And hmmmmm, I am not sure about this.

Do the particles take the player speed into account? Otherwise it would look unrealistic as soon you move.
Can the particles be culled by nodes? Otherwise it will rain indoors. ;) Interaction with nodes is my biggest concern. The illusion of snow/rain could be quickly noticed by the player.
Are there other use cases besides weather?

I appreciate your experiments with this, but I don't know if this will be actually useful. I agree that a FAST method for weather particles is very desirable.

**Pull request closed without merging**